proposals/NNNN-parameter-packs.md

```diff
@@ -0,0 +1,837 @@
+
+For example, say that `x` and `y` are both value parameter packs and `T` is a type parameter pack, and consider the pack expansion expression `repeat foo(each x, (each T).self, (repeat each y))`. This expression captures both the value parameter pack `x` and type parameter pack `T`, but it does not capture `y`, because `y` is captured by the inner pack expansion expression `repeat each y`. Additionally, if `x` has the type `Foo<U, (repeat each V)>`, then our expression captures `U`, but not `V`, because again, `V` is captured by the inner pack expansion type `repeat each V`.
+
+**Typing rules:** For a pack expansion expression to be well-typed, two conditions must hold:
+
+1. The types of all value parameter packs captured by a pack expansion expression must be related via same-shape requirements.
+
+2. After replacing all value parameter packs with non-pack parameters that have equivalent types, the pattern expression must be well-typed.
+
+Assuming the above hold, the type of a pack expansion expression is defined to be the pack expansion type whose pattern type is the type of the pattern expression.
+
+**Evaluation semantics:** At runtime, each value (or type) parameter pack receives a value (or type) pack, which is a concrete list of values (or types). The same-shape requirements guarantee that all value (and type) packs have the same length, call it `N`. The evaluation semantics are that for each successive `i` such that `0 ≤ i < N`, the pattern expression is evaluated after substituting each occurrence of a value (or type) parameter pack with the `i`th element of the value (or type) pack. The evaluation proceeds from left to right according to the usual evaluation order, and the list of results from each evaluation forms the argument list for the parent expression.
+
+For example, pack expansion expressions can be used to forward value parameter packs to other functions:
+
+```swift
+func tuplify<each T>(_ t: repeat each T) -> (repeat each T) {
+  return (repeat each t)
+}
+
+func forward<each U>(u: repeat each U) {
+  let _ = tuplify(repeat each u) //  T := {repeat each U}
+  let _ = tuplify(repeat each u, 10) // T := {repeat each U, Int}
+  let _ = tuplify(repeat each u, repeat each u) // T := {repeat each U, repeat each U}
+  let _ = tuplify(repeat [each u]) // T := {repeat Array<each U>}
+}
+```
+
+### Overload resolution
+
+Generic functions can be overloaded by the "pack-ness" of their type parameters. For example, a function can have two overloads where one accepts a scalar type parameter and the other accepts a type parameter pack:
+
+```swift
+func overload<T>(_: T) {}
+func overload<each T>(_: repeat each T) {}
+```
+
+If both overloads match a given call, e.g. `overload(1)`, the call is ambiguous. Similarly, two generic functions where one accepts a non-pack variadic parameter and the other accepts a type parameter pack:
```

Comment:
What should happen for calls like `overload()` if the first overload has default value i.e. -`overload<T>(_: T = 42)`? I think we need a special rule here to define ranking between variadic with zero values and defaulted generic parameter?

Reply:
@xedin Would you mind sharing those thoughts over in the [review discussion](https://forums.swift.org/t/se-0393-value-and-type-parameter-packs/63859)?